packages/macros/src/ember-addon-main.ts

```diff
@@ -48,16 +50,23 @@ export = {
 
     appInstance.import('vendor/embroider-macros-test-support.js', { type: 'test' });
 
-    // When we're used inside the traditional ember-cli build pipeline without
-    // Embroider, we unfortunately need to hook into here uncleanly because we
-    // need to delineate the point in time after which writing macro config is
-    // forbidden and consuming it becomes allowed. There's no existing hook with
-    // that timing.
     const originalToTree = appInstance.toTree;
-    appInstance.toTree = function (...args) {
-      macrosConfig.finalize();
-      return originalToTree.apply(appInstance, args);
-    };
+
+    // as macrosConfig is 1:1 with the appInstance, if
+    // the toTree method is already wrapped, we don't need
+    // to wrap it again
+    if (!wrappedTrees.has(originalToTree)) {
```

Comment:
Does it have to be a `Set`? Couldn't it be just a simple boolean flag? I.e. we will only be called with a single `appInstance`, right?

Reply:
That sounds right... I wasn't sure if this could somehow get hit with multiple apps in the same process. Maybe during tests?

Reply:
Updated without using a Set. It works for me locally this way.